_posts/2021-09-17-the-joy-of-slow-programming.md

```diff
@@ -0,0 +1,160 @@
+dropped to zero I did `rm -rf the-project` and told my friend: "Well, I guess
+that I need to think about it more in the next days".
+
+### New approach - take it slow
+
+The result of this thinking were a few guiding ideas that I decided on:
+1. use ["boring" technologies](/blog/2018/11/22/boring-ruby-code/) that I already
+know - Ruby, Nokogiri, Rails, and Heroku.
+2. no more than one hour of work on the project per day
+3. focus on delivering the smallest possible thing that actually provides some
+value to my friend
+4. strive for simplicity from the technical perspective and not perfection
+
+I thought about them as defining my new mindset of "slow programming". And this
+is how [Neue Wohnung](https://github.com/adamniedzielski/neue_wohnung) was born.
+If you look at the commits in the reverse order you can see that I started with
+the implementation, added tests soon, but not immediately, and configured CI
+after I had a running version in production.
+
+I was working at a slow, steady pace, trying to enjoy the every moment of it.
+As I removed the pressure of time from myself, I could actually enjoy the
+little details like a [new Rubocop rule that I didn't know before](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/ConstantDefinitionInBlock)
+or read up on [libraries that I was considering to use](https://github.com/bblimke/webmock).
+
+After a few days of working I had the first version and could actually start
+sending notifications to my friend. This meant that I was receiving feedback
+early and could adjust the application based on it, without making too many
+false assumptions.
+
+Was the fast "hacking evening" with Python a wasted time? No, not really.
+Thanks to this session I learned that Telegram is a messaging app where it's
+extremely easy to send messages programmatically and how to create my own
+Telegram bot. From the time perspective I think about it as prototyping session,
+although during the actual session I didn't have this prototyping mindset.
+
+### Next boring steps
+
+The scraper was working in production (I quickly set up Heroku) and my friend was
+receiving new apartment offers every day. The next step was very boring - add
+support for more companies. It was a tedious and repetitive activity to write
+this code, but I was never doing it for more than an hour per day and at the
+same time immediately getting the business value out of it - more and more
+apartment offers were reaching my friend - and this kept me motivated.
+
+I always started by copy pasting the previous scraper and the tests for it.
+The code was repetitive but I avoided reaching for any early abstractions.
+The classes were simple and could fit in a single screen. The actual business
+logic - getting information from the HTML using Nokogiri and regular expressions
+- was unique to each company website; with very few common elements.
+
+This approach paid off as soon as I found the first website loading data
```

Comment:
I don't understand how the logical connection you're trying to make between this paragraph and the previous one - in what way does "discovering that data was loaded dynamically" mean that your approach was good?

Reply:
The connection is following: there are no early abstractions so it's easier to adapt the code to a new situation where we parse JSON and not HTML. I'll see how I can make that clearer.